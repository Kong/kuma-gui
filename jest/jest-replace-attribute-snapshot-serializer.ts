// Imports the appropriate plugin type from the library Jest uses for snapshot serialization.
import { NewPlugin } from 'pretty-format'

const DEFAULT_UUID = 'aaaabbbb-cccc-dddd-eeee-ffffffffffff'

function hasUuidAttribute(element: Element, attribute: string): boolean {
  const value = element.getAttribute(attribute)

  return value !== null && /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/.test(value)
}

const processedValues = new Set()

export function replaceAttributesSnapshotSerializer(
  attributes: string[],
  uuid = DEFAULT_UUID,
): NewPlugin {
  return {
    /**
     * Matches elements which havenâ€™t been processed by this serializer before.
     */
    test: (value) => value instanceof Element && !processedValues.has(value)
      ? attributes.some((attribute) => hasUuidAttribute(value, attribute))
      : false,

    /**
     * Replaces the attribute values of an element with a static value.
     */
    serialize: (value: Element, config, indentation, depth, refs, printer) => {
      for (const attribute of attributes) {
        if (hasUuidAttribute(value, attribute)) {
          value.setAttribute(attribute, uuid)
        }
      }

      processedValues.add(value)

      return printer(value, config, indentation, depth, refs)
    },
  }
}
