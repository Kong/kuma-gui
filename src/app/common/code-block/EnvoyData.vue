<template>
  <div>
    <DataSource
      v-slot="{ data, error, refresh }: EnvoyDataSource"
      :src="props.src"
    >
      <ErrorBlock
        v-if="error"
        :error="error"
      />

      <LoadingBlock v-else-if="data === undefined" />

      <EmptyBlock v-else-if="data === ''" />

      <CodeBlock
        v-else
        language="json"
        :code="typeof data === 'string' ? data : JSON.stringify(data, null, 2)"
        is-searchable
        :query="props.query"
        :is-filter-mode="props.isFilterMode"
        :is-reg-exp-mode="props.isRegExpMode"
        @query-change="emit('query-change', $event)"
        @filter-mode-change="emit('filter-mode-change', $event)"
        @reg-exp-mode-change="emit('reg-exp-mode-change', $event)"
      >
        <template #primary-actions>
          <KButton
            appearance="primary"
            @click="refresh"
          >
            <RefreshIcon />

            Refresh
          </KButton>
        </template>
      </CodeBlock>
    </DataSource>
  </div>
</template>

<script lang="ts" setup>
import { RefreshIcon } from '@kong/icons'

import CodeBlock from './CodeBlock.vue'
import EmptyBlock from '@/app/common/EmptyBlock.vue'
import ErrorBlock from '@/app/common/ErrorBlock.vue'
import LoadingBlock from '@/app/common/LoadingBlock.vue'
import { EnvoyDataSource } from '@/app/zones/sources'

const props = withDefaults(defineProps<{
  resource: string
  src: string
  query?: string
  isFilterMode?: boolean
  isRegExpMode?: boolean
}>(), {
  query: '',
  isFilterMode: false,
  isRegExpMode: false,
})

const emit = defineEmits<{
  (event: 'query-change', query: string): void
  (event: 'filter-mode-change', isFilterMode: boolean): void
  (event: 'reg-exp-mode-change', isRegExpMode: boolean): void
}>()
</script>
