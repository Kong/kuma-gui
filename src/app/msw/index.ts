import { setupWorker } from 'msw/browser'

import { token } from '@/services/utils'
import type { Token, ServiceDefinition } from '@/services/utils'
import type { SetupWorker } from 'msw/browser'

type Handler = Parameters<typeof setupWorker>[0]
type Passthrough = (req: Request) => string | string[] | void

type Msw = ReturnType<SetupWorker['start']>

const $ = {
  msw: token<Msw>('msw'),
  mswHandlers: token('msw.handlers'),
  mswPassthroughs: token('msw.passthroughs'),
}
export const TOKENS = $
const passthroughExtension = /\.(vue|ts|js|json)(\?.*)?$/
export const services = (_: Record<string, Token>): ServiceDefinition[] => {
  return [
    [token('msw.passthrough.msw'), {
      service: () => [(req: Request) => {
        const { pathname, href } = new URL(req.url)
        if (
          !pathname.startsWith('/node_modules') &&
         !pathname.startsWith('/src/assets/') &&
         !passthroughExtension.test(href)
        ) {
          return `Found an unhandled ${req.method} request to ${href}`
        }
      }],
      labels: [
        $.mswPassthroughs,
      ],
    },
    ],
    [$.msw, {
      service: (handlers: Handler[] = [], passthroughs: Passthrough[]) => {
        return setupWorker(...handlers).start({
          quiet: true,
          onUnhandledRequest(req: Request) {
            passthroughs.reduce<string[]>((prev, item) => prev.concat(item(req) ?? []), []).forEach(item => {
              console.warn(item)
            })
          },
        })
      },
      arguments: [
        $.mswHandlers,
        $.mswPassthroughs,
      ],
    }],
  ]
}
