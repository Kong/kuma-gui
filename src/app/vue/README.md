---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

