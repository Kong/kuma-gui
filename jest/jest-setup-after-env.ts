import { afterAll, afterEach, beforeAll, beforeEach, expect, jest } from '@jest/globals'

// Polyfills `window.fetch` for Jest because it runs in a Node environment where fetch isn’t available. It initially looked like this would change with Node.js 18, but that is not so.
import 'isomorphic-fetch'

import { config } from '@vue/test-utils'

import { replaceAttributesSnapshotSerializer } from './jest-replace-attribute-snapshot-serializer'
import { createRouter } from '../src/router/router'
import { TOKENS, get, container, set, injected } from '../src/services'
import { TOKENS as COMPONENT_TOKENS } from '../src/components'
import { setupMockServer } from '../src/api/setupMockServer'
import { rest, MockedRequest as Request } from 'msw'
import Env from '@/services/env/Env'

type MockFunction = (_opts: Record<string, unknown>, cb: (req: Request, resp: Record <string, any>) => Record<string, unknown>) => void

// jest can't import this module properly due to transpiling issues
// mock this out with a blank element
jest.mock('vue-github-button', () => ({ template: '<span />' }))

/**
 * Adds the application’s router to vue test utils. This way tests don’t have to set-up a new router instance on their own.
 */
const router = createRouter(get(TOKENS.routes))
config.global.plugins.push(router)

/**
 * Adds the application’s Vuex store to vue test utils. This way tests don’t have to set-up a new store instance on their own.
 */
config.global.plugins.push([get(TOKENS.store), get(TOKENS.storeKey)])

/**
 * Kongponents uses generated UUIDs for several attribute values.
 * This breaks the project’s snapshot tests since they’re based on fully-mounted components
 * which also includes those from external sources like Kongponents.
 *
 * In order to stabilize the tests which otherwise fail because the attribute values are different every run,
 * we use a custom snapshot serializer to replace those attribute values with one fixed value.
 */
expect.addSnapshotSerializer(replaceAttributesSnapshotSerializer([
  'id',
  'aria-describedby',
  'aria-labelledby',
  'aria-controls',
  'data-tableid',
]))

const server = setupMockServer(import.meta.env.VITE_KUMA_API_SERVER_URL)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

// unless we actually use COMPONENT_TOKENS it won't actually get executed
// probably due to tree shaking/rollup import ordering. This mixed with
// container capturing/restoring to make our tests isolated means that
// potentially we can capture an empty container before all the tokens are set,
// then set the TOKENS/fill the container during a test then the container can
// get restored to empty whilst we still have TOKENS with now non-existent
// services accessing TOKENS before we do anything means we set the TOKENS and
// fill the container with the default services i.e. before we capture if we
// ever make a test mocking utility to mock out components (similar to
// withVersion) will will then use COMPONENT_TOKENS here also, which means we
// can remove the following line
beforeAll((_ = COMPONENT_TOKENS) => {})
//
beforeEach(() => container.capture?.())
afterEach(() => container.restore?.())

// add a utility to easily setup/mock out API endpoints
const re = /\+/g

const useMock = (url: string, response: Record<string, unknown>):MockFunction => {
  return (_opts, cb) => {
    server.use(
      rest.get(`${import.meta.env.VITE_KUMA_API_SERVER_URL.slice(0, -1)}${url.replace(re, '\\+')}`, (req, res, ctx) => {
        return res(ctx.json(cb(req, JSON.parse(JSON.stringify(response)))))
      }),
    )
  }
}

export const withVersion = (v: string) => {
  class TestEnv extends Env {
    var(...rest: Parameters<Env['var']>) {
      const key = rest[0]
      if (key === 'KUMA_VERSION') {
        return v
      }
      return super.var(...rest)
    }
  }
  set(TOKENS.Env, TestEnv)
  injected(TestEnv, TOKENS.EnvVars)
}
export { router, server, useMock }
