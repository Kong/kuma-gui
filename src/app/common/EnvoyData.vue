<template>
  <div class="envoy-data">
    <div class="envoy-data-actions">
      <KButton
        :disabled="isLoading"
        appearance="primary"
        icon="redo"
        data-testid="envoy-data-refresh-button"
        @click="fetchContent"
      >
        Refresh
      </KButton>
    </div>

    <StatusInfo
      :has-error="error !== null"
      :is-loading="isLoading"
      :error="error"
    >
      <CodeBlock
        :id="`code-block-${props.dataPath}`"
        language="json"
        :code="code"
        is-searchable
        :query-key="props.queryKey ?? `code-block-${props.dataPath}`"
      />
    </StatusInfo>
  </div>
</template>

<script lang="ts" setup>
import { onMounted, PropType, ref, watch } from 'vue'
import { KButton } from '@kong/kongponents'

import CodeBlock from './CodeBlock.vue'
import StatusInfo from './StatusInfo.vue'
import { useKumaApi } from '@/utilities'

const kumaApi = useKumaApi()

const props = defineProps({
  dataPath: {
    type: String as PropType<'xds' | 'stats' | 'clusters'>,
    required: true,
  },

  queryKey: {
    type: String,
    required: false,
    default: null,
  },

  mesh: {
    type: String,
    required: false,
    default: '',
  },

  dppName: {
    type: String,
    required: false,
    default: '',
  },

  zoneIngressName: {
    type: String,
    required: false,
    default: '',
  },

  zoneEgressName: {
    type: String,
    required: false,
    default: '',
  },
})

const isLoading = ref(true)
const error = ref<Error | null>(null)
const code = ref('')

watch(() => props.dppName, function () {
  fetchContent()
})
watch(() => props.zoneIngressName, function () {
  fetchContent()
})
watch(() => props.zoneEgressName, function () {
  fetchContent()
})

onMounted(function () {
  fetchContent()
})

async function fetchContent() {
  error.value = null
  isLoading.value = true

  try {
    let content = ''

    if (props.mesh !== '' && props.dppName !== '') {
      content = await kumaApi.getDataplaneData({
        dataPath: props.dataPath,
        mesh: props.mesh,
        dppName: props.dppName,
      })
    } else if (props.zoneIngressName !== '') {
      content = await kumaApi.getZoneIngressData({
        dataPath: props.dataPath,
        zoneIngressName: props.zoneIngressName,
      })
    } else if (props.zoneEgressName !== '') {
      content = await kumaApi.getZoneEgressData({
        dataPath: props.dataPath,
        zoneEgressName: props.zoneEgressName,
      })
    }

    code.value = typeof content === 'string' ? content : JSON.stringify(content, null, 2)
  } catch (err) {
    if (err instanceof Error) {
      error.value = err
    } else {
      console.error(err)
    }
  } finally {
    isLoading.value = false
  }
}
</script>

<style lang="scss" scoped>
.envoy-data {
  padding: var(--spacing-md);
}

.envoy-data-actions {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  margin-bottom: var(--spacing-md);
}
</style>
