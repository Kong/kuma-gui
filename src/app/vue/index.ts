import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
