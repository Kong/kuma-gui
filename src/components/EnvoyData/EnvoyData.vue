<template>
  <StatusInfo
    class="envoy-data"
    :has-error="error !== undefined"
    :is-loading="isLoading"
    :error="error"
  >
    <CodeBlock
      :id="`code-block-${dataPath}`"
      language="json"
      :code="code"
      is-searchable
      :query-key="queryKey ?? `code-block-${dataPath}`"
    />
  </StatusInfo>
</template>

<script lang="ts" setup>
import { onMounted, PropType, ref, watch } from 'vue'

import Kuma from '@/services/kuma'
import CodeBlock from '@/app/common/CodeBlock.vue'
import StatusInfo from '@/components/Utils/StatusInfo.vue'

const props = defineProps({
  dataPath: {
    type: String as PropType<'xds' | 'stats' | 'clusters'>,
    required: true,
  },

  queryKey: {
    type: String,
    required: false,
    default: null,
  },

  mesh: {
    type: String,
    required: false,
    default: '',
  },

  dppName: {
    type: String,
    required: false,
    default: '',
  },

  zoneIngressName: {
    type: String,
    required: false,
    default: '',
  },

  zoneEgressName: {
    type: String,
    required: false,
    default: '',
  },
})

const isLoading = ref(true)
const error = ref<Error | undefined>(undefined)
const code = ref('')

watch(() => props.dppName, function () {
  fetchContent()
})
watch(() => props.zoneIngressName, function () {
  fetchContent()
})
watch(() => props.zoneEgressName, function () {
  fetchContent()
})

onMounted(function () {
  fetchContent()
})

async function fetchContent() {
  error.value = undefined
  isLoading.value = true

  try {
    let content = ''

    if (props.mesh !== '' && props.dppName !== '') {
      content = await Kuma.getDataplaneData({
        dataPath: props.dataPath,
        mesh: props.mesh,
        dppName: props.dppName,
      })
    } else if (props.zoneIngressName !== '') {
      content = await Kuma.getZoneIngressData({
        dataPath: props.dataPath,
        zoneIngressName: props.zoneIngressName,
      })
    } else if (props.zoneEgressName !== '') {
      content = await Kuma.getZoneEgressData({
        dataPath: props.dataPath,
        zoneEgressName: props.zoneEgressName,
      })
    }

    code.value = typeof content === 'string' ? content : JSON.stringify(content, null, 2)
  } catch (err) {
    if (err instanceof Error) {
      error.value = err
    } else {
      console.error(err)
    }
  } finally {
    isLoading.value = false
  }
}
</script>

<style lang="scss">
.code-container .k-card-body {
  position: relative;
}
</style>

<style lang="scss" scoped>
.envoy-data {
  padding: var(--spacing-md);
}

.copy-button {
  position: absolute;
  top: var(--spacing-md);
  right: var(--spacing-md);
  display: block;
}
</style>
