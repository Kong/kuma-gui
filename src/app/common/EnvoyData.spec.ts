import { render, screen } from '@testing-library/vue'
import { rest } from 'msw'

import EnvoyData from './EnvoyData.vue'
import { server } from '@/../jest/jest-setup-after-env'

function renderComponent(props = {}) {
  return render(EnvoyData, {
    props: {
      dataPath: 'clusters',
      ...props,
    },
  })
}

describe('EnvoyData.vue', () => {
  it('renders snapshot', async () => {
    server.use(
      rest.get(import.meta.env.VITE_KUMA_API_SERVER_URL + 'meshes/:mesh/dataplanes/:dataplaneName/clusters', (req, res, ctx) =>
        res(ctx.status(200), ctx.json('')),
      ),
    )

    const { container } = renderComponent({
      mesh: 'foo',
      dppName: 'dataplane-test-456',
    })

    expect(container).toMatchSnapshot()
  })

  it('renders loading', () => {
    server.use(
      rest.get(import.meta.env.VITE_KUMA_API_SERVER_URL + 'meshes/:mesh/dataplanes/:dataplaneName/clusters', (req, res, ctx) =>
        res(ctx.status(200), ctx.json('')),
      ),
    )

    renderComponent({
      mesh: 'foo',
      dppName: 'dataplane-test-456',
    })

    expect(screen.getByTestId('loading-block')).toBeInTheDocument()
  })

  it('renders error', async () => {
    server.use(
      rest.get(import.meta.env.VITE_KUMA_API_SERVER_URL + 'meshes/:mesh/dataplanes/:dataplaneName/clusters', (req, res, ctx) =>
        res(ctx.status(500), ctx.json('')),
      ),
    )

    renderComponent({
      mesh: 'test-mesh',
      dppName: 'dataplane-test-456',
    })

    expect((await screen.findAllByText(/An error has occurred while trying to load this data./))[0]).toBeInTheDocument()
  })
})
