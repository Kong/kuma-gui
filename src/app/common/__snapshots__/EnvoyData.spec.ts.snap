// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvoyData.vue renders snapshot 1`] = `
<div
  class="envoy-data"
>
  <div
    class="envoy-data-actions"
  >
    <button
      class="k-button medium rounded primary"
      data-testid="envoy-data-refresh-button"
      disabled=""
      type="button"
    >
      
      <span
        class="k-button-icon kong-icon kong-icon-redo"
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 17 14"
          width="17"
          xmlns="http://www.w3.org/2000/svg"
        >
          
  
          <title>
            Redo
          </title>
          
  
          <path
            clip-rule="evenodd"
            d="M12.466 12.06a6.868 6.868 0 112.196-5.676l.85-.85a.847.847 0 011.201-.017c.327.327.315.869-.017 1.2L14.233 9.18c-.342.342-.88.35-1.208.023L10.654 6.83c-.324-.324-.314-.859.028-1.2.34-.34.881-.348 1.2-.029l.847.846a4.938 4.938 0 10-1.63 4.245l1.367 1.367z"
            fill="#A3BBCC"
            fill-rule="evenodd"
          />
          

        </svg>
        

      </span>
      
      
       Refresh 
      
      <!---->
    </button>
  </div>
  <div>
    <section
      class="empty-state-wrapper"
      data-testid="loading-block"
    >
      <div
        class="empty-state-title"
      >
        <!---->
        <div
          class="k-empty-state-title-header mt-4 mb-4"
        >
          
          <span
            class="mb-3 kong-icon kong-icon-spinner"
          >
            <svg
              fill="none"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              
  
              <title>
                Loading
              </title>
              
  
              <g>
                
    
                <path
                  d="M5 10a5 5 0 1 0 5-5"
                  fill="none"
                  stroke="#A3BBCC"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                
  
              </g>
              

            </svg>
            

          </span>
          
          <p>
            Loading data â€¦
          </p>
          
          
        </div>
      </div>
      <div
        class="empty-state-content"
      >
        <div
          class="k-empty-state-message mb-6"
        >
          
          
        </div>
        <div
          class="k-empty-state-cta"
        >
          
          <!---->
          
        </div>
      </div>
    </section>
  </div>
</div>
`;
